### PR TITLE
Database Interop

### PR DESCRIPTION
**Pull request by bstpierr**
Note that many Database integration tests appear to be failing on the master branch as well as in my branch at the moment. I'm not sure if this is expected, of if I need a valid Google-Info.plist file to make the tests work beyond local fakes.

Register Database with Interop:

  + Add AuthInterop dependency
  + Have Database register with Interop component container
  + Add weak dependency on Auth
  + Cleaned up imports

Use Interop for Auth token fetching:

  + Use macro to get auth component
  + Change `getTokenForcingRefresh:withCallback:` call to use Interop component version of Auth
  + Clean up imports

Implement necessary protocols:

  + Added FIRComponentRegistrant and FIRDatabaseNilProtocol to FIRDatabase

Clean up Database tests:

  + Removed all unnecessary header files
  + Ran style.sh across all the tests

Cleaned Up project file:

  Some header removals were missed in the last commit.

Sorted some test files

Fix Database Fake App Tests:

  + Added container property
  + Added an Auth fake in the container
  + Added the Shared utils files to the necessary targets
  + Added a missing XCTest reference in a test that didn't compile for me

### CODE REVIEW

**Comment by bstpierr** [comment]
It appears as though the Database integration tests is a KI #1809.

**Comment by ryanwilson** [comment]
It looks like `clang-format` changed a large number of lines that aren't related to this change - do we know why this is? Was the code originally compliant with the `clang-format` settings we're using or are those files typically ignored? It's a bit hard to review the actual changes at the moment with a lines changed count of (+15,348 −13,427).

**Comment by paulb777** [comment]
tldr: Remove style changes from PR and pass integration tests

- It's fine to run small style fixes in a PR, but they should not be squashed
- In general, there's no reason to squash commits on a PR since standard practice is to squash on merge
- Big style changes should be their own PR.
- Database doesn't do style because there are some header file ordering requirements that were too much to fix when we first started running `clang-format`.

- The integration tests should still run with a valid `GoogleServices-Info.plist`. See instructions at https://github.com/firebase/firebase-ios-sdk#running-sample-apps.
- The flaky tests were ifdef'd out

- `git reflog` can help to recover the squashed commit history

**Comment by bstpierr** [comment]
Yeah, once I uploaded the PR last night and saw the line count I figured I'd have to remove the style changes so I `reflog`'ed back to a previous detached head state and rewound the squash, but it was getting late to fix it. I'll fix this branch up and ping the comments once I think it's ready to be looked at again.

**Comment by bstpierr** [comment]
Okay, this should be closer to the minimal changes now (+92 −54). I merged more times than I needed to, but I was able to get all the changes in without the major re-style.

I'm running the integration tests locally now, with a valid plist.

**Comment by paulb777** [comment]
Based on travis run, looks like there are some build issues to resolve as well - https://travis-ci.org/firebase/firebase-ios-sdk/jobs/432514473

File: Firebase/Database/Api/FIRDatabase.m [hunk 1]
```diff
@@ -69,6 +77,23 @@ + (void)load {
...
+  [FIRComponentContainer registerAsComponentRegistrant:self];
+}
+
++ (NSArray<FIRComponent *> *)componentsToRegister {
+    FIRDependency *auth =
+        [FIRDependency dependencyWithProtocol:@protocol(FIRAuthInterop)
+                                   isRequired:NO];
+    FIRComponentCreationBlock creationBlock =
+        ^id _Nullable(FIRComponentContainer *container, BOOL *isCacheable) {
+        return [self databaseForApp:container.app];
```
Comment (ryanwilson) [inline Firebase/Database/Api/FIRDatabase.m]: Let's chat about this when I get in, but I think we should follow the Storage PR https://github.com/firebase/firebase-ios-sdk/pull/1660 which will also move the instance management to the container. It'd also help keep this class a bit cleaner without much knowledge of the component system and just inject the `auth` instance as needed.

File: Firebase/Database/Login/FAuthTokenProvider.m [hunk 2]
```diff
@@ -15,9 +15,14 @@
  */
 
 #import "FAuthTokenProvider.h"
-#import "FUtilities.h"
+
+#import <FirebaseAuthInterop/FIRAuthInterop.h>
 #import <FirebaseCore/FIRAppInternal.h>
+#import <FirebaseCore/FIRComponent.h>
```
Comment (ryanwilson) [inline Firebase/Database/Login/FAuthTokenProvider.m]: Instead of this class using the component system directly, we could pass in the `auth` instance in the constructor instead of the `app`. This would simplify testing greatly too since we could just use fakes instead of having to mock `FIRApp` at all.

The notification can still be used since `Auth` passes in `self` as the object for the notification, so we can compare `notification.object == self.auth` instead of the app and the app from the notification's dictionary.

File: Firebase/Database/Api/FIRDatabase.m (hunk 1 above)
Comment (bstpierr) [reply Firebase/Database/Api/FIRDatabase.m]: I can make a component class like `FIRStorageComponent` and remove as many references to the `FIRApp` as possible from the initializers in `FIRDatabase`, only passing around the `auth` instance instead.

Is this what you mean?

Comment (bstpierr) [reply Firebase/Database/Api/FIRDatabase.m]: This has all moved to the new `FIRDatabaseProvider` class.

File: Firebase/Database/Login/FAuthTokenProvider.m (hunk 2 above)
Comment (bstpierr) [reply Firebase/Database/Login/FAuthTokenProvider.m]: Done.

File: Firebase/Database/Api/FIRDatabaseComponent.h [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+@class FIRApp;
+@class FIRDatabase;
+
+NS_ASSUME_NONNULL_BEGIN
+
+/// This protocol is used in the interop registration process to register an instance provider for
+/// individual FIRApps.
+@protocol FIRDatabaseProvider
+
+/// Default method for creating a Storage instance.
```
Comment (ryanwilson) [inline Firebase/Database/Api/FIRDatabaseComponent.h]: `s/Storage/Database/` 😄

File: Firebase/Database/Api/FIRDatabaseComponent.m [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+                      instantiationTiming:FIRInstantiationTimingLazy
+                             dependencies:@[ authDep ]
+                            creationBlock:creationBlock];
+  return @[ databaseProvider ];
+}
+
+#pragma mark - FIRDatabaseProvider Conformance
+
+- (FIRDatabase *)databaseForApp:(FIRApp *)app URL:(NSString *)url {
+  id<FIRAuthInterop> auth = FIR_COMPONENT(FIRAuthInterop, self.app.container);
```
Comment (ryanwilson) [inline Firebase/Database/Api/FIRDatabaseComponent.m]: Auth should be fetched here, but this is where we should be using the database initializer (mentioned in a previous comment in `FIRDatabase.m`).

File: Firebase/Database/Login/FAuthTokenProvider.m [hunk 5]
```diff
@@ -64,45 +67,52 @@ - (void) dealloc {
...
-- (void) fetchTokenForcingRefresh:(BOOL)forceRefresh withCallback:(fbt_void_nsstring_nserror)callback {
-    // TODO: Don't fetch token if there is no current user
-    [self.app getTokenForcingRefresh:forceRefresh withCallback:^(NSString * _Nullable token, NSError * _Nullable error) {
+- (void)fetchTokenForcingRefresh:(BOOL)forceRefresh
+                     withCallback:(fbt_void_nsstring_nserror)callback {
+  if (self.auth == nil) {
+    // Signal that Auth is not available by returning nil.
+    callback(nil, nil);
+  } else {
+    [self.auth getTokenForcingRefresh:forceRefresh
```
Comment (ryanwilson) [inline Firebase/Database/Login/FAuthTokenProvider.m]: Based on the `TODO` above, we should also check if the `getUserID` call returns nil and not call `getTokenForcingRefresh:withCallback`.

File: FirebaseDatabase.podspec [hunk 6]
```diff
@@ -32,6 +32,7 @@ Simplify your iOS development, grow your user base, and monetize more effectivel
   s.libraries = ['c++', 'icucore']
   s.frameworks = ['CFNetwork', 'Security', 'SystemConfiguration']
   s.dependency 'leveldb-library', '~> 1.18'
+  s.dependency 'FirebaseAuthInterop', '~> 1.0'
   s.dependency 'FirebaseCore', '~> 5.0'
```
Comment (ryanwilson) [inline FirebaseDatabase.podspec]: `FirebaseCore` also needs to go to `5.1` and above - we missed this before in the other PRs. `5.1` is where the component system was introduced.

File: Firebase/Database/Api/FIRDatabase.m [hunk 7]
```diff
@@ -140,7 +143,7 @@ + (FIRDatabase *)databaseForApp:(FIRApp *)app URL:(NSString *)url {
     } else if (![databaseUrl.path isEqualToString:@""] && ![databaseUrl.path isEqualToString:@"/"]) {
         [NSException raise:@"InvalidDatabaseURL" format:@"Configured Database URL '%@' is invalid. It should point "
             "to the root of a Firebase Database but it includes a path: %@",databaseUrl, databaseUrl.path];
-  }
+    }
 
     FIRDatabaseDictionary *instances = [self instances];
```
[the remark this answers is not in the document]
Comment (bstpierr) [reply Firebase/Database/Api/FIRDatabase.m]: Done.

File: Firebase/Database/Api/FIRDatabase.m [hunk 8]
```diff
@@ -154,7 +157,9 @@ + (FIRDatabase *)databaseForApp:(FIRApp *)app URL:(NSString *)url {
         FParsedUrl *parsedUrl = [FUtilities parseUrl:databaseUrl.absoluteString];
         FIRDatabase *database = urlInstanceMap[parsedUrl.repoInfo];
         if (!database) {
-            id<FAuthTokenProvider> authTokenProvider = [FAuthTokenProvider authTokenProviderForApp:app];
```
[the remark this answers is not in the document]
Comment (bstpierr) [reply Firebase/Database/Api/FIRDatabase.m]: Done.

File: Firebase/Database/Api/FIRDatabaseComponent.h (hunk 3 above)
Comment (bstpierr) [reply Firebase/Database/Api/FIRDatabaseComponent.h]: Done :)

File: Firebase/Database/Api/FIRDatabaseComponent.m (hunk 4 above)
Comment (bstpierr) [reply Firebase/Database/Api/FIRDatabaseComponent.m]: Done.

File: FirebaseDatabase.podspec (hunk 6 above)
Comment (bstpierr) [reply FirebaseDatabase.podspec]: Done.

File: Firebase/Database/Login/FAuthTokenProvider.m (hunk 5 above)
Comment (bstpierr) [reply Firebase/Database/Login/FAuthTokenProvider.m]: Done.

**Comment by bstpierr** [comment]
There are still integration tests that are failing that I need to change. I'll ping the PR again when they're fixed.

File: Firebase/Database/Api/FIRDatabase.m [hunk 9]
```diff
@@ -38,142 +41,27 @@ @interface FIRDatabase ()
...
 + (FIRDatabase *)databaseForApp:(FIRApp *)app {
-    if (app == nil) {
-        [NSException raise:@"InvalidFIRApp" format:@"nil FIRApp instance passed to databaseForApp."];
-    }
-
     return [FIRDatabase databaseForApp:app URL:app.options.databaseURL];
 }
 
 + (FIRDatabase *)databaseForApp:(FIRApp *)app URL:(NSString *)url {
-    if (app == nil) {
```
Comment (ryanwilson) [inline Firebase/Database/Api/FIRDatabase.m]: I think it's still okay to validate the app here, since otherwise the `app.container` will be nil and won't actually log any errors.

File: Firebase/Database/Api/FIRDatabase.m [hunk 10]
```diff
@@ -38,142 +41,27 @@ @interface FIRDatabase ()
...
-    static dispatch_once_t pred = 0;
-    static FIRDatabaseDictionary *instances;
-    dispatch_once(&pred, ^{
-        instances = [NSMutableDictionary dictionary];
-    });
-    return instances;
-}
-
 + (FIRDatabase *)database {
-    if (![FIRApp isDefaultAppConfigured]) {
```
Comment (ryanwilson) [inline Firebase/Database/Api/FIRDatabase.m]: This is still worth keeping since `[FIRApp defaultApp]` will log an error anyways and it's a developer error.

File: Firebase/Database/Api/FIRDatabaseComponent.m [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+  }
+  return self;
+}
+
+#pragma mark - Lifecycle
+
++ (void)load {
+  [FIRComponentContainer registerAsComponentRegistrant:self];
+  NSNotificationCenter *center = [NSNotificationCenter defaultCenter];
+  [center addObserverForName:kFIRAppDeleteNotification
```
Comment (ryanwilson) [inline Firebase/Database/Api/FIRDatabaseComponent.m]: There's a new protocol, `FIRComponentLifecycleMaintainer` that has an `appWillBeDeleted:` call that can replace this. The component container will loop through cached instances when an app is deleted and if the instance conforms to `FIRComponentLifecycleMaintainer` will call `appWillBeDeleted:`.

File: Firebase/Database/Api/FIRDatabase.m [hunk 12]
```diff
@@ -38,142 +41,27 @@ @interface FIRDatabase ()
...
-            database = [[FIRDatabase alloc] initWithApp:app
-                                               repoInfo:parsedUrl.repoInfo
-                                                 config:config];
-            urlInstanceMap[parsedUrl.repoInfo] = database;
-        }
-
-        return database;
-    }
+  id<FIRDatabaseProvider> provider = FIR_COMPONENT(FIRDatabaseProvider, app.container);
+  return [provider databaseForApp:app URL: url];
```
Comment (ryanwilson) [inline Firebase/Database/Api/FIRDatabase.m]: Nit: extra space after `URL:`

File: Firebase/Database/Api/FIRDatabaseComponent.h [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+@class FIRDatabase;
+
+NS_ASSUME_NONNULL_BEGIN
+
+/// This protocol is used in the interop registration process to register an instance provider for
+/// individual FIRApps.
+@protocol FIRDatabaseProvider
+
+/// Gets the instance of FIRDatabase for the default FIRApp.
+- (FIRDatabase *)database;
```
Comment (ryanwilson) [inline Firebase/Database/Api/FIRDatabaseComponent.h]: It looks like the only method we actually need here is `databaseForApp:URL:` - the rest aren't called.

File: Firebase/Database/Api/FIRDatabaseComponent.m [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
+    if (!urlInstanceMap) {
+      urlInstanceMap = [NSMutableDictionary dictionary];
+      instances[app.name] = urlInstanceMap;
+    }
+
+    FParsedUrl *parsedUrl = [FUtilities parseUrl:databaseUrl.absoluteString];
+    FIRDatabase *database = urlInstanceMap[parsedUrl.repoInfo];
+    if (!database) {
+      id<FAuthTokenProvider> authTokenProvider =
+      [FAuthTokenProvider authTokenProviderWithAuthInterop:
```
Comment (ryanwilson) [inline Firebase/Database/Api/FIRDatabaseComponent.m]: Nit: indentation

File: Example/Database/Tests/Integration/FIRAuthTests.m [hunk 15]
```diff
@@ -40,7 +43,9 @@ - (void)tearDown {
 
 - (void)testListensAndAuthRaceCondition {
     [FIRDatabase setLoggingEnabled:YES];
-    id<FAuthTokenProvider> tokenProvider = [FAuthTokenProvider authTokenProviderForApp:[FIRApp defaultApp]];
+    id<FAuthTokenProvider> tokenProvider =
+        [FAuthTokenProvider authTokenProviderWithAuthInterop:
+            FIR_COMPONENT(FIRAuthInterop, [FIRApp defaultApp].container)];
```
Comment (ryanwilson) [inline Example/Database/Tests/Integration/FIRAuthTests.m]: It would be nice to avoid the default app here, since we've seen it cause some issues with flakiness previously. Luckily, the container shouldn't even be needed: there's an auth interop fake [`FIRAuthInteropFake`](https://github.com/firebase/firebase-ios-sdk/blob/master/Example/Shared/FIRAuthInteropFake.h) that simulates Auth functionality - take a look at the [storage examples here](https://github.com/firebase/firebase-ios-sdk/blob/7e5befc30356604a9f324bb74664a970d148f493/Example/Storage/Tests/Unit/FIRStorageTokenAuthorizerTests.m#L35).

File: Example/Database/Tests/Helpers/FTestHelpers.m [hunk 16]
```diff
@@ -55,7 +60,9 @@ + (NSArray*) getRandomNodes:(int)num persistence:(BOOL)persistence {
 
     NSMutableArray *refs = (persistence) ? persistenceRefs : noPersistenceRefs;
 
-    id<FAuthTokenProvider> authTokenProvider = [FAuthTokenProvider authTokenProviderForApp:[FIRApp defaultApp]];
+    id<FAuthTokenProvider> authTokenProvider =
+        [FAuthTokenProvider authTokenProviderWithAuthInterop:
```
Comment (ryanwilson) [inline Example/Database/Tests/Helpers/FTestHelpers.m]: Same as the other comment - 
"It would be nice to avoid the default app here, since we've seen it cause some issues with flakiness previously. Luckily, the container shouldn't even be needed: there's an auth interop fake [`FIRAuthInteropFake`](https://github.com/firebase/firebase-ios-sdk/blob/master/Example/Shared/FIRAuthInteropFake.h) that simulates Auth functionality - take a look at the storage examples [here](https://github.com/firebase/firebase-ios-sdk/blob/7e5befc30356604a9f324bb74664a970d148f493/Example/Storage/Tests/Unit/FIRStorageTokenAuthorizerTests.m#L35)."

Comment (bstpierr) [reply Example/Database/Tests/Helpers/FTestHelpers.m]: Done

File: Example/Database/Tests/Integration/FIRAuthTests.m (hunk 15 above)
Comment (bstpierr) [reply Example/Database/Tests/Integration/FIRAuthTests.m]: Done

File: Firebase/Database/Api/FIRDatabase.m (hunk 10 above)
Comment (bstpierr) [reply Firebase/Database/Api/FIRDatabase.m]: Done

File: Firebase/Database/Api/FIRDatabase.m (hunk 9 above)
Comment (bstpierr) [reply Firebase/Database/Api/FIRDatabase.m]: Done

File: Firebase/Database/Api/FIRDatabase.m (hunk 12 above)
Comment (bstpierr) [reply Firebase/Database/Api/FIRDatabase.m]: Done

File: Firebase/Database/Api/FIRDatabaseComponent.h (hunk 13 above)
Comment (bstpierr) [reply Firebase/Database/Api/FIRDatabaseComponent.h]: Done

File: Firebase/Database/Api/FIRDatabaseComponent.m (hunk 11 above)
Comment (bstpierr) [reply Firebase/Database/Api/FIRDatabaseComponent.m]: Done

**Comment by paulb777** [comment]
Looks like the Auth unit test build is broken on travis - https://travis-ci.org/firebase/firebase-ios-sdk/jobs/452003504

File: Firebase/Database/Api/FIRDatabaseComponent.m [hunk 17]
```diff
@@ -0,0 +1,190 @@
...
+#import <FirebaseCore/FIROptions.h>
+
+NS_ASSUME_NONNULL_BEGIN
+
+/** A NSMutableDictionary of FirebaseApp name and FRepoInfo to FirebaseDatabase instance. */
+typedef NSMutableDictionary<NSString *, NSMutableDictionary<FRepoInfo *, FIRDatabase *> *>
+    FIRDatabaseDictionary;
+
+@interface FIRDatabase ()
+@property (nonatomic, strong) FRepoInfo *repoInfo;
```
[the remark this answers is not in the document]
Comment (bstpierr) [reply Firebase/Database/Api/FIRDatabaseComponent.m]: I moved them over to FIRDatabase_Private.h

File: Firebase/Database/Api/FIRDatabaseConfig.m [hunk 18]
```diff
@@ -106,9 +110,11 @@ + (FIRDatabaseConfig *)configForName:(NSString *)name {
     });
     @synchronized(sessionConfigs) {
         if (!sessionConfigs[name]) {
-            id<FAuthTokenProvider> authTokenProvider = [FAuthTokenProvider authTokenProviderForApp:[FIRApp defaultApp]];
-            sessionConfigs[name] = [[FIRDatabaseConfig alloc] initWithSessionIdentifier:name
-                                                                      authTokenProvider:authTokenProvider];
+            id<FAuthTokenProvider> authTokenProvider =
+                [FAuthTokenProvider authTokenProviderWithAuthInterop:
+                    FIR_COMPONENT(FIRAuthInterop, [FIRApp defaultApp].container)];
```
[the remark this answers is not in the document]
Comment (bstpierr) [reply Firebase/Database/Api/FIRDatabaseConfig.m]: I just went ahead and removed the code as per the existing TODOs and updated the tests.

File: Firebase/Database/Login/FAuthTokenProvider.h [hunk 19]
```diff
@@ -19,19 +19,20 @@
...
 
-- (void) listenForTokenChanges:(fbt_void_nsstring)listener;
+- (void)listenForTokenChanges:(fbt_void_nsstring)listener;
 
 @end
 
 @interface FAuthTokenProvider : NSObject
 
-+ (id<FAuthTokenProvider>) authTokenProviderForApp:(FIRApp *)app;
++ (id<FAuthTokenProvider>)authTokenProviderWithAuthInterop:(id<FIRAuthInterop>)authInterop;
```
[the remark this answers is not in the document]
Comment (bstpierr) [reply Firebase/Database/Login/FAuthTokenProvider.h]: Done.

File: Firebase/Database/Api/FIRDatabaseComponent.m [hunk 20]
```diff
@@ -0,0 +1,182 @@
...
+    if (!urlInstanceMap) {
+      urlInstanceMap = [NSMutableDictionary dictionary];
+      instances[app.name] = urlInstanceMap;
+    }
+
+    FParsedUrl *parsedUrl = [FUtilities parseUrl:databaseUrl.absoluteString];
+    FIRDatabase *database = urlInstanceMap[parsedUrl.repoInfo];
+    if (!database) {
+      id<FAuthTokenProvider> authTokenProvider =
+      [FAuthTokenProvider authTokenProviderWithAuth:
```
Comment (ryanwilson) [inline Firebase/Database/Api/FIRDatabaseComponent.m]: Nit: spacing

File: Firebase/Database/Api/FIRDatabaseConfig.m [hunk 21]
```diff
@@ -14,11 +14,15 @@
  * limitations under the License.
  */
 
-#import <FirebaseCore/FIRApp.h>
 #import "FIRDatabaseConfig.h"
+
+#import <FirebaseAuthInterop/FIRAuthInterop.h>
```
Comment (ryanwilson) [inline Firebase/Database/Api/FIRDatabaseConfig.m]: These three imports should be removed.

File: Firebase/Database/Login/FAuthTokenProvider.m [hunk 22]
```diff
@@ -15,27 +15,31 @@
  */
 
 #import "FAuthTokenProvider.h"
-#import "FUtilities.h"
+
+#import <FirebaseAuthInterop/FIRAuthInterop.h>
 #import <FirebaseCore/FIRAppInternal.h>
+#import <FirebaseCore/FIRComponent.h>
```
Comment (ryanwilson) [inline Firebase/Database/Login/FAuthTokenProvider.m]: I believe this can be removed, along with the App import, only keeping the AuthInterop import.

File: Example/Database/Tests/Helpers/FIRFakeApp.m [hunk 23]
```diff
@@ -16,33 +16,49 @@
...
+          return [[FIRAuthInteropFake alloc] initWithToken:nil userID:nil error:nil];
+        };
+    FIRComponentCreationBlock databaseBlock =
+        ^id _Nullable(FIRComponentContainer *container, BOOL *isCacheable) {
+          return [[FIRDatabaseComponent alloc] initWithApp:container.app];
+        };
+    NSDictionary<NSString *, FIRComponentCreationBlock> *components =
+        @{NSStringFromProtocol(@protocol(FIRAuthInterop)) : authBlock,
+          NSStringFromProtocol(@protocol(FIRDatabaseProvider)) : databaseBlock};
+    _container = [[FIRComponentContainer alloc] initWithApp:(FIRApp *)self components:components];
```
[the remark this answers is not in the document]
Comment (bstpierr) [reply Example/Database/Tests/Helpers/FIRFakeApp.m]: `initWithApp:` sets the newly initialized instance's app property, so `_container.app` should be set on init.

Is there something else that needs set that I'm missing?

File: Example/Database/Tests/Helpers/FTestHelpers.m [hunk 24]
```diff
@@ -42,6 +48,18 @@ + (NSTimeInterval) waitUntil:(BOOL (^)())predicate timeout:(NSTimeInterval)secon
     return (finish - start);
 }
 
++ (FIRDatabaseConfig *) defaultConfig {
+  return [self configForName:@"default"];
+}
+
++ (FIRDatabaseConfig *) configForName:(NSString *)name {
+  id<FIRAuthInterop> auth =
+  [[FIRAuthInteropFake alloc] initWithToken:nil userID:nil error:nil];
```
[the remark this answers is not in the document]
Comment (bstpierr) [reply Example/Database/Tests/Helpers/FTestHelpers.m]: Done.

File: Example/Database/Tests/Integration/FIRAuthTests.m [hunk 25]
```diff
@@ -40,10 +44,14 @@ - (void)tearDown {
...
     [FIRDatabase setLoggingEnabled:YES];
-    id<FAuthTokenProvider> tokenProvider = [FAuthTokenProvider authTokenProviderForApp:[FIRApp defaultApp]];
-
-    FIRDatabaseConfig *config = [FIRDatabaseConfig configForName:@"testWritesRestoredAfterAuth"];
-    config.authTokenProvider = tokenProvider;
+    FIRAuthInteropFake *auth =
+        [[FIRAuthInteropFake alloc] initWithToken:nil userID:nil error:nil];
+    id<FAuthTokenProvider> authTokenProvider =
+        [FAuthTokenProvider authTokenProviderWithAuth:auth];
+            FIR_COMPONENT(FIRAuthInterop, [FIRApp defaultApp].container);
```
[the remark this answers is not in the document]
Comment (bstpierr) [reply Example/Database/Tests/Integration/FIRAuthTests.m]: Whoops!

File: Firebase/Database/Api/FIRDatabase.m [hunk 26]
```diff
@@ -16,98 +16,48 @@
...
-    return instances;
-}
-
 + (FIRDatabase *)database {
     if (![FIRApp isDefaultAppConfigured]) {
         [NSException raise:@"FIRAppNotConfigured"
                     format:@"Failed to get default Firebase Database instance. Must call `[FIRApp "
-                           @"configure]` (`FirebaseApp.configure()` in Swift) before using "
-                           @"Firebase Database."];
+         @"configure]` (`FirebaseApp.configure()` in Swift) before using "
```
[the remark this answers is not in the document]
Comment (bstpierr) [reply Firebase/Database/Api/FIRDatabase.m]: Done.

File: Firebase/Database/Api/FIRDatabase.m [hunk 27]
```diff
@@ -16,98 +16,48 @@
...
 + (FIRDatabase *)databaseWithURL:(NSString *)url {
     FIRApp *app = [FIRApp defaultApp];
-    if (app == nil) {
-        [NSException raise:@"FIRAppNotConfigured"
-                    format:@"Failed to get default Firebase Database instance. "
-                           @"Must call `[FIRApp configure]` (`FirebaseApp.configure()` in Swift) "
-                           @"before using Firebase Database."];
+        if (app == nil) {
+            [NSException raise:@"FIRAppNotConfigured"
+                        format:@"Failed to get default Firebase Database instance. "
```
[the remark this answers is not in the document]
Comment (bstpierr) [reply Firebase/Database/Api/FIRDatabase.m]: Done.

File: Firebase/Database/Api/FIRDatabaseComponent.m [hunk 28]
```diff
@@ -0,0 +1,182 @@
...
+     "Specify a valid DatabaseURL within FIRApp or from your databaseForApp:URL: call.", databaseUrl];
+  } else if (![databaseUrl.path isEqualToString:@""] && ![databaseUrl.path isEqualToString:@"/"]) {
+    [NSException raise:@"InvalidDatabaseURL" format:@"Configured Database URL '%@' is invalid. It should point "
+     "to the root of a Firebase Database but it includes a path: %@",databaseUrl, databaseUrl.path];
+  }
+
+  FIRDatabaseDictionary *instances = [FIRDatabaseComponent instances];
+  @synchronized (instances) {
+    NSMutableDictionary<FRepoInfo *, FIRDatabase *> *urlInstanceMap =
+    instances[app.name];
```
[the remark this answers is not in the document]
Comment (bstpierr) [reply Firebase/Database/Api/FIRDatabaseComponent.m]: Done.

File: Firebase/Database/Api/FIRDatabaseComponent.m [hunk 29]
```diff
@@ -0,0 +1,182 @@
...
+      };
+  FIRComponent *databaseProvider =
+      [FIRComponent componentWithProtocol:@protocol(FIRDatabaseProvider)
+                      instantiationTiming:FIRInstantiationTimingLazy
+                             dependencies:@[ authDep ]
+                            creationBlock:creationBlock];
+  return @[ databaseProvider ];
+}
+
+#pragma mar - Instance management.
```
[the remark this answers is not in the document]
Comment (bstpierr) [reply Firebase/Database/Api/FIRDatabaseComponent.m]: Done.

File: Firebase/Database/Api/FIRDatabaseConfig.m (hunk 21 above)
Comment (bstpierr) [reply Firebase/Database/Api/FIRDatabaseConfig.m]: Done.

File: Firebase/Database/Login/FAuthTokenProvider.m (hunk 22 above)
Comment (bstpierr) [reply Firebase/Database/Login/FAuthTokenProvider.m]: Actually, the AppInternal is needed for the `FIRAuthStateDidChangeInternalNotification` declaration.

Comment (bstpierr) [reply Firebase/Database/Login/FAuthTokenProvider.m]: Actually, the AppInternal is needed for the

File: Firebase/Database/Api/FIRDatabaseComponent.m [hunk 30]
```diff
@@ -0,0 +1,182 @@
...
+ * A static NSMutableDictionary of FirebaseApp name and FRepoInfo to
+ * FirebaseDatabase instance. To ensure thread-safety, it should only be
+ * accessed in databaseForApp:URL:, which is synchronized.
+ *
+ * TODO: This serves a duplicate purpose as RepoManager.  We should clean up.
+ * TODO: We should maybe be conscious of leaks and make this a weak map or
+ * similar but we have a lot of work to do to allow FirebaseDatabase/Repo etc.
+ * to be GC'd.
+ */
++ (FIRDatabaseDictionary *)instances {
```
[the remark this answers is not in the document]
Comment (bstpierr) [reply Firebase/Database/Api/FIRDatabaseComponent.m]: Done.

**Comment by bstpierr** [comment]
Fixed everything you mentioned unless I left the comment unresolved.

File: Firebase/Database/Api/FIRDatabaseComponent.m [hunk 31]
```diff
@@ -35,6 +35,7 @@
     FIRDatabaseDictionary;
 
 @interface FIRDatabaseComponent () <FIRComponentLifecycleMaintainer, FIRComponentRegistrant>
+@property (nonatomic) FIRDatabaseDictionary *instances;
```
Comment (ryanwilson) [inline Firebase/Database/Api/FIRDatabaseComponent.m]: This doesn't need to be `FIRDatabaseDictionary` anymore, since it's supposed to be keyed on app name. Since this can only be instantiated with a single app, it should just be `NSMutableDictionary<FRepoInfo *, FIRDatabase *> *`. `FIRDatabaseDictionary` can be removed entirely.

File: Firebase/Database/Api/FIRDatabaseComponent.m [hunk 32]
```diff
@@ -75,33 +77,14 @@ + (void)load {
...
 
 - (void)appWillBeDeleted:(FIRApp *)app {
   NSString *appName = app.name;
   if (appName == nil) {
     return;
   }
-  FIRDatabaseDictionary* instances = [FIRDatabaseComponent instances];
+  FIRDatabaseDictionary* instances = [self instances];
   @synchronized (instances) {
     NSMutableDictionary<FRepoInfo *, FIRDatabase *> *databaseInstances = instances[appName];
```
Comment (ryanwilson) [inline Firebase/Database/Api/FIRDatabaseComponent.m]: This needs to   change based on the above change. Also, it doesn't look like we actually use FRepoInfo as a key here, so maybe it  makes sense to use something else (URL?)

**Comment by bstpierr** [comment]
This should be good to look at again.

File: Example/Firebase.xcodeproj/project.pbxproj [hunk 33]
```diff
@@ -5105,7 +5109,7 @@
 				CODE_SIGN_ENTITLEMENTS = Messaging/Sample/iOS/Messaging_Example.entitlements;
 				"CODE_SIGN_IDENTITY[sdk=iphoneos*]" = "iPhone Developer";
 				DEBUG_INFORMATION_FORMAT = dwarf;
-				DEVELOPMENT_TEAM = EQHXZ8M8AV;
+				DEVELOPMENT_TEAM = "";
```
[the remark this answers is not in the document]
Comment (paulb777) [reply Example/Firebase.xcodeproj/project.pbxproj]: LGTM

File: Firebase/Database/Api/FIRDatabaseComponent.m [hunk 34]
```diff
@@ -0,0 +1,154 @@
...
+                      instantiationTiming:FIRInstantiationTimingLazy
+                             dependencies:@[ authDep ]
+                            creationBlock:creationBlock];
+  return @[ databaseProvider ];
+}
+
+#pragma mark - Instance management.
+
+- (void)appWillBeDeleted:(FIRApp *)app {
+  NSString *appName = app.name;
```
Comment (schmidt-sebastian) [inline Firebase/Database/Api/FIRDatabaseComponent.m]: Question: Why do we need to pass the app here? Is it always the same app as the one passed to the constructor?
